core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -1018,7 +1025,9 @@ private[spark] class BlockManager(
           try {
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.dispose()
+            if (!level.useOffHeap) {
```

Comment:
Similarly, aren't we deciding whether we should dispose `bytesToReplicate` based on the 'target' `StorageLevel`? Shouldn't we make that decision based on whether `bytesToReplicate` were stored off-heap or not?